### ISSUE TITLE
jormungandr - probable memory leak

### ISSUE THREAD

**Issue by rinor** [input-output-hk/jormungandr#1835]
**Describe the bug**
The node appears to leak memory in some places.

Will analyse just one of those cases that may be plaguing the trusted nodes the most.

---
- Will be calculating based on **Proportional Set Size (_PSS_)** since is supposed to represent the amount of memory used by libraries and the node and includes only physical memory usage, not the memory that has been swapped out to disk
- Will keep a _static state of the chain_ during all tests. So the main node is started without trusted peers and secrets so it just sits upon the storage data (around mid-epoch 68).
- Client and Server are _**not** on the same machine_.
---
 Single client case:
---
When we start a node (**A**), once it loads the chain from storage and/or network (bootstraps), it settles down to a certain amount of memory depending on the chain length. 

Node (**A**):  (data in KB) 
```json
  PID  Swap      USS      PSS      RSS 
15830     0  1303136  1303358  1305624
```
Lets start 1 new node (**B**), with _**empty storage**_ that will bootstrap the whole chain, including block0, from our node (**A**). Once the connections is estabilished and block streaming starts the memory usage increases within a couple of seconds with **_248MB_** (this is not a fixed value):

Node (**A**):  (data in KB) 
```json
  PID  Swap      USS      PSS      RSS 
15830     0  1557360  1557569  1559848 
```
Once bootstrap completed, we shutdown node (**B**) and check the memory on
Node (**A**):  (data in KB) 
```json
  PID  Swap      USS      PSS      RSS 
15830     0  1557616  1557813  1560104 
``` 
Now if you repeat the procedure, by starting a new node (or the same node (**B**) after deleting the storage) that will bootstrap from node (**A**) , the memory usage keeps increasing. The amount is not fixed, but has some patterns. 

The allocated memory won't be ever released.

---
 Concurrent clients case:
---
The same as the first scenario but this time will fire 4 nodes (**B**) with empty storage that will bootstrap from node (**A**).

When we start a node (**A**), once it loads the chain from storage and/or network (bootstraps), it settles down to a certain amount of memory depending on the chain length. 

Node (**A**):  (data in KB) 
```json
  PID  Swap      USS      PSS      RSS 
21584     0  1302832  1302997  1305236 
```

Lets start 4 nodes (**B**), with _**empty storage**_ that will bootstrap the whole chain, including block0, from our node (**A**). Once the connections is estabilished and block streaming starts the memory usage increases within a couple of seconds with **_923MB_** (this is not a fixed value but approx case 1 x 4):

Node (**A**):  (data in KB) 
```json
  PID  Swap      USS      PSS      RSS 
21584     0  2248472  2248637  2250876
```
Once bootstrap completed, we shutdown all 4 nodes (**B**) and check the memory on
Node (**A**):  (data in KB) 
```json
  PID  Swap      USS      PSS      RSS 
21584     0  2248980  2249151  2251380 
``` 
The allocated memory won't be ever released.

---

**Mandatory Information**
1. `jormungandr 0.8.11 (master-a209f7eb, release, linux [x86_64]) - [rustc 1.41.0 (5e1a79984 2020-01-27)]`
2. `#1 SMP Debian 5.4.13-1~bpo10+1 (2020-02-07) x86_64 GNU/Linux`

**To Reproduce**
- explained on description

**Expected behavior**
Unused memory should be released/reused.

**Additional context**

**Comment by vincenthz** [comment]
I don't think we have any memory going unaccounted, even less likely a "memory leak" in the real sense of the term. It's very possible that while jormungandr is not using the memory anymore, the memory allocator (jemalloc) doesn't see a reason to free the pages to the system, and decide to hold on for it's own use later.

I think the first action item, is to add the jemalloc_ctl stats in the log / rest api (node stats) to get some handle on the problem; for example, we have this (for platform that supports it): https://github.com/input-output-hk/chain-libs/blob/master/imhamt/examples/memdump/unix.rs

The second is to monitor the state of the system, for example dumping the stats of the ledger would be useful too with `Ledger::info()`

**Comment by eugene-babichenko** [comment]
Why not try using [`std::alloc::System`](https://doc.rust-lang.org/beta/std/alloc/struct.System.html)? Will it significantly hit the performance?

At least we can compare the results with `jemalloc`.

**Comment by michaeljfazio** [comment]
I did some recon on this issue. With a memory profiler I managed to extract the following:

![Screenshot from 2020-03-15 01-59-29](https://user-images.githubusercontent.com/1845178/76687883-f024e780-6662-11ea-853c-703de522ec34.png)

Top memory hungry call graphs (collapsed):

![image](https://user-images.githubusercontent.com/1845178/76688013-fd8ea180-6663-11ea-967e-2f54f6f8d675.png)

Top four expanded call graphs:

![image](https://user-images.githubusercontent.com/1845178/76687958-9113a280-6663-11ea-9af6-a62c4d9ad541.png)

![image](https://user-images.githubusercontent.com/1845178/76687970-a688cc80-6663-11ea-9558-d2dd2824fd37.png)

![image](https://user-images.githubusercontent.com/1845178/76687986-c02a1400-6663-11ea-92a4-9efd44c38ebc.png)

![image](https://user-images.githubusercontent.com/1845178/76688000-d6d06b00-6663-11ea-8dd9-485256013b7b.png)

This profile only ran for about 30 minutes. I can repeat if necessary for longer duration. Hopefully this will help someone isolate the issue.

Otherwise, might be worth doing a binary search over commits to track down the commit which introduced the problem using `git bisect`. Between the tip of master and 0.8.9 release, there are only 7 steps using bisect.

**Comment by noemx** [comment]
Observation on Linux with Jormungandr 0.8.18 running as leader for 10 days with default connection configuration:

After 1 day: 2.86GB RAM used.
After 9 days: 3.2GB RAM used.
This means daily increase of 43 MB.

RAM used after Jormungandr shutdown: 130MB

![2020-04-22 16_43_31-Window](https://user-images.githubusercontent.com/180145/79997264-aa590b80-84b9-11ea-9fb3-06eed4fe9e08.png)
